python/cudf/cudf/core/udf/pipeline.py

```diff
@@ -234,8 +301,27 @@ def _kernel(retval, input_col_0, input_col_1, offset_0, offset_1, size):
     return kernel_template.format(**d)
 
 
+def _check_return_type(ty):
+    """
+    In almost every case, get_udf_return_type will throw a typing error
+    if the user tries to use a field in the row containing a dtype that
+    is not supported. It does this by simply "not finding" overloads of
+    any operators for the ancillary dtype we type that field as. But it
+    is defeated by the following edge case:
+
+    def f(row):
+        return row[<bad dtype key>]
+
+    In this case numba is happy to return MaskedType(<bad dtype key>),
+    and won't throw because no operators were ever used. This function
+    explicitly checks for that case.
+    """
+    if isinstance(ty, Dummy):
+        raise TypeError(str(ty))
+
+
 @annotate("UDF COMPILATION", color="darkgreen", domain="cudf_python")
-def compile_or_get(df, f):
+def compile_or_get(md):
```

Comment:
As noted above in the discussion of `_FrameJitMetadata`, I think the previous function signature was clearer in its intent and separation of types -- though it wouldn't hurt to use the name `func` instead of `f`. The docstring below may need to be updated if this change is kept to accept a single instance of `_FrameJitMetadata` as a parameter.

Reply:
Went back to two args here passing the function and the metadata explicitly. What's better though, instantiating the `_FrameJitMetadata` inside `Frame._apply` and passing it to `compile_or_get`, or passing `frame` to `compile_or_get` and creating the `_FrameJitMetadata` inside it?